nexus/src/db/lookup.rs

```diff
@@ -0,0 +1,723 @@
+                    [<$pc:lower _lookup_by_name_no_authz>](
+                        opctx,
+                        datastore,
+                        name
+                    ).await?;
+                opctx.authorize(authz::Action::Read, &authz_child).await?;
+                Ok((authz_child, db_child))
+            }
+
+            impl LookupNoauthz for $pc<'_> {
+                type LookupType = authz::$pc;
+
+                fn lookup(
+                    &self,
+                ) -> BoxFuture<'_, LookupResult<Self::LookupType>> {
+                    async {
+                        let lookup = self.lookup_root();
+                        let opctx = &lookup.opctx;
+                        let datastore = lookup.datastore;
+                        match self.key {
+                            Key::Name(_, name) => {
+                                let (rv, _) =
+                                    [<$pc:lower _lookup_by_name_no_authz>](
+                                        opctx,
+                                        datastore,
+                                        name
+                                    ).await?;
+                                Ok(rv)
+                            }
+                            Key::Id(_, id) => {
+                                let (rv, _) =
+                                    [<$pc:lower _lookup_by_id_no_authz>](
+                                        opctx,
+                                        datastore,
+                                        id
+                                    ).await?;
+                                Ok(rv)
+                            }
+                        }
+                    }.boxed()
+                }
+            }
+
+            impl Fetch for $pc<'_> {
+                type FetchType = (authz::$pc, model::$pc);
+
+                fn fetch(&self) -> BoxFuture<'_, LookupResult<Self::FetchType>> {
+                    async {
+                        let lookup = self.lookup_root();
+                        let opctx = &lookup.opctx;
+                        let datastore = lookup.datastore;
+                        match self.key {
+                            Key::Name(_, name) => {
+                                [<$pc:lower _fetch_by_name>](
+                                    opctx,
+                                    datastore,
+                                    name
+                                ).await
+                            }
+                            Key::Id(_, id) => {
+                                [<$pc:lower _fetch_by_id>](
+                                    opctx,
+                                    datastore,
+                                    id
+                                ).await
+                            }
+                        }
+                    }
+                    .boxed()
+                }
+            }
+        }
+    };
+}
+
+macro_rules! define_lookup_with_parent {
```

Comment:
Out of curiosity, why does this need to be a macro? I'm comparing this to something like https://github.com/oxidecomputer/omicron/blob/4279d237678d27697c9c4215714a9d8ae1a1f639/nexus/src/db/collection_insert.rs#L67-L75 , which opts to define the "parent -> child" relationship as a trait.

Although going with the trait-based approach requires explicitly enumerating the columns being acted upon, it has some advantages:

- The trait only defines a single link in the lookup tree, and does not require us to re-state the "entire tree" at any point
- (Somewhat subjective) I think it's easier to read / understand than macros, when possible

Reply:
Mentioned elsewhere: I went down this path and gave up:
https://github.com/oxidecomputer/omicron/pull/798#issuecomment-1075774779